### PR TITLE
Feature asynch io layers

### PR DESCRIPTION
**Pull request by bvanessen**
Add support for double buffering in the I/O layers and fetching data asynchronously.

### CODE REVIEW

File: include/lbann/data_readers/data_reader.hpp [hunk 1]
```diff
@@ -740,13 +736,15 @@ class generic_data_reader : public lbann_image_preprocessor {
...
+  bool fetch_data_block(CPUMat& X, El::Int thread_index, El::Int mb_size, El::Matrix<El::Int>& indices_fetched, thread_pool& io_thread_pool);
+
   /**
    * Fetch a single sample into a matrix.
    * @param X The matrix to load data into.
    * @param data_id The index of the datum to fetch.
    * @param mb_idx The index within the mini-batch.
    */
-  virtual bool fetch_datum(CPUMat& X, int data_id, int mb_idx, int tid) {
+  virtual bool fetch_datum(CPUMat& X, int data_id, int mb_idx, thread_pool& io_thread_pool) {
```
Comment (timmoon10) [inline include/lbann/data_readers/data_reader.hpp]: I don't think it's necessary to make the thread pool available here. This function is overridden in every derived class and they only need to know the thread index (e.g. to access per-thread buffers). It's best to concentrate complexity in the base class.

File: src/data_readers/data_reader_mesh.cpp [hunk 2]
```diff
@@ -66,7 +66,8 @@ void mesh_reader::load() {
   select_subset_of_data();
 }
 
-bool mesh_reader::fetch_datum(CPUMat& X, int data_id, int mb_idx, int tid) {
+bool mesh_reader::fetch_datum(CPUMat& X, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_mesh.cpp]: If this is not used, we should just delete the line.

File: src/data_readers/data_reader_mesh.cpp [hunk 3]
```diff
@@ -82,7 +83,8 @@ bool mesh_reader::fetch_datum(CPUMat& X, int data_id, int mb_idx, int tid) {
   return true;
 }
 
-bool mesh_reader::fetch_response(CPUMat& Y, int data_id, int mb_idx, int tid) {
+bool mesh_reader::fetch_response(CPUMat& Y, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_mesh.cpp]: Ditto.

File: src/data_readers/data_reader_mnist.cpp [hunk 4]
```diff
@@ -48,7 +48,8 @@ void mnist_reader::set_defaults() {
   m_num_labels=10;
 }
 
-bool mnist_reader::fetch_datum(CPUMat& X, int data_id, int mb_idx, int tid) {
+bool mnist_reader::fetch_datum(CPUMat& X, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_mnist.cpp]: Ditto.

File: src/data_readers/data_reader_mnist_siamese.cpp [hunk 5]
```diff
@@ -222,7 +217,8 @@ bool data_reader_mnist_siamese::fetch_datum(CPUMat& X, std::pair<int, int> data_
 }
 
 
-bool data_reader_mnist_siamese::fetch_label(CPUMat& Y, std::pair<int, int> data_id, int mb_idx, int tid) {
+bool data_reader_mnist_siamese::fetch_label(CPUMat& Y, std::pair<int, int> data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_mnist_siamese.cpp]: Ditto.

File: src/data_readers/data_reader_moving_mnist.cpp [hunk 6]
```diff
@@ -62,8 +62,9 @@ int moving_mnist_reader::get_linearized_label_size() const {
   return get_num_labels();
 }
 
-bool moving_mnist_reader::fetch_datum(CPUMat& X, int data_id, int col, int tid) {
+bool moving_mnist_reader::fetch_datum(CPUMat& X, int data_id, int col, thread_pool& io_thread_pool) {
 
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_moving_mnist.cpp]: Ditto.

File: src/data_readers/data_reader_moving_mnist.cpp [hunk 7]
```diff
@@ -201,8 +201,9 @@ bool moving_mnist_reader::fetch_datum(CPUMat& X, int data_id, int col, int tid)
   return true;
 }
 
-bool moving_mnist_reader::fetch_label(CPUMat& Y, int data_id, int col, int tid) {
+bool moving_mnist_reader::fetch_label(CPUMat& Y, int data_id, int col, thread_pool& io_thread_pool) {
 
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_moving_mnist.cpp]: Ditto.

File: src/data_readers/data_reader_multi_images.cpp [hunk 8]
```diff
@@ -140,10 +141,10 @@ std::vector<data_reader_multi_images::sample_t> data_reader_multi_images::get_im
   std::vector<sample_t> ret;
   ret.reserve(m_mini_batch_size);
 
-  for (El::Int i = 0; i < m_indices_fetched_per_mb.Height(); ++i) {
-    El::Int index = m_indices_fetched_per_mb.Get(i, 0);
-    ret.push_back(m_image_list[index]);
-  }
+  // for (El::Int i = 0; i < m_indices_fetched_per_mb.Height(); ++i) {
```
Comment (ndryden) [inline src/data_readers/data_reader_multi_images.cpp]: Ditto.

File: src/data_readers/data_reader_numpy.cpp [hunk 9]
```diff
@@ -130,7 +130,8 @@ void numpy_reader::load() {
   select_subset_of_data();
 }
 
-bool numpy_reader::fetch_datum(Mat& X, int data_id, int mb_idx, int tid) {
+bool numpy_reader::fetch_datum(Mat& X, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_numpy.cpp]: Ditto.

File: src/data_readers/data_reader_numpy.cpp [hunk 10]
```diff
@@ -149,7 +150,8 @@ bool numpy_reader::fetch_datum(Mat& X, int data_id, int mb_idx, int tid) {
   return true;
 }
 
-bool numpy_reader::fetch_label(Mat& Y, int data_id, int mb_idx, int tid) {
+bool numpy_reader::fetch_label(Mat& Y, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_numpy.cpp]: Ditto.

File: src/data_readers/data_reader_numpy.cpp [hunk 11]
```diff
@@ -165,7 +167,8 @@ bool numpy_reader::fetch_label(Mat& Y, int data_id, int mb_idx, int tid) {
   return true;
 }
 
-bool numpy_reader::fetch_response(Mat& Y, int data_id, int mb_idx, int tid) {
+bool numpy_reader::fetch_response(Mat& Y, int data_id, int mb_idx, thread_pool& io_thread_pool) {
+  //  int tid = io_thread_pool.get_local_thread_id();
```
Comment (ndryden) [inline src/data_readers/data_reader_numpy.cpp]: Ditto.

File: src/data_readers/data_reader_triplet.cpp [hunk 12]
```diff
@@ -125,10 +126,10 @@ std::vector<data_reader_triplet::sample_t> data_reader_triplet::get_image_list_o
   std::vector<sample_t> ret;
   ret.reserve(m_mini_batch_size);
 
-  for (El::Int i = 0; i < m_indices_fetched_per_mb.Height(); ++i) {
-    El::Int index = m_indices_fetched_per_mb.Get(i, 0);
-    ret.emplace_back(m_samples.get_sample(index));
-  }
+  // for (El::Int i = 0; i < m_indices_fetched_per_mb.Height(); ++i) {
```
Comment (ndryden) [inline src/data_readers/data_reader_triplet.cpp]: Ditto.

File: src/utils/threads/thread_utils.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/utils/threads/thread_utils.hpp"
+#include <thread>
+#include <omp.h>
+
+namespace lbann {
+
+int num_free_cores_per_process(const lbann_comm *comm) {
```
Comment (ndryden) [inline src/utils/threads/thread_utils.cpp]: Note that this excludes any threads our communication libraries (Aluminum, MPI, NCCL, ...) or the like may launch.

File: include/lbann/base.hpp [hunk 14]
```diff
@@ -163,11 +163,17 @@ static bool __attribute__((used)) endsWith(const std::string mainStr, const std:
...
 
 /// Print the dimensions and name of a Elemental matrix
-static void __attribute__((used)) _display_matrix(ElMat *m, const char *name) {
+static void __attribute__((used)) _display_matrix(AbsDistMat *m, const char *name) {
   std::cout << "DISPLAY MATRIX: " << name << " = " << m->Height() << " x " << m->Width() << std::endl;
 }
 #define DISPLAY_MATRIX(x) _display_matrix(x, #x);
 
+/// Print the dimensions and name of a Elemental matrix
+static void __attribute__((used)) _display_local_matrix(AbsMat *m, const char *name) {
```
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/base.hpp]: Fixed.

File: include/lbann/data_readers/data_reader.hpp (hunk 1 above)
Comment (bvanessen) [reply include/lbann/data_readers/data_reader.hpp]: We cannot pass the thread id into the function because we do not bind the execution context with the function when the lambda is enqueued.  As a result we need to dynamically get the thread id.

Comment (bvanessen) [reply include/lbann/data_readers/data_reader.hpp]: The thread_pool is now a member variable.

**Comment by bvanessen** [comment]
@ndryden I updated the data readers to keep the thread pool as a member function and to no longer pass it as a parameter to fetch functions.

File: include/lbann/layers/io/input/generic_input_layer.hpp [hunk 15]
```diff
@@ -848,21 +928,26 @@ class generic_input_layer : public io_layer {
   }
 
  protected:
+  std::vector<generic_io_buffer*> m_io_buffers;
```
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/layers/io/input/generic_input_layer.hpp]: I would like to defer making this change until I clean up the rest of the I/O system and get rid of the partitioned and distributed I/O buffers.  At that point, it will be easier to change this to a unique pointer.

File: include/lbann/io/data_buffers/partitioned_io_buffer.hpp [hunk 16]
```diff
@@ -41,26 +41,61 @@ class partitioned_io_buffer : public generic_io_buffer {
...
-  void set_std_matrix_view(El::Int cur_mini_batch_size, int idx) override {}
-  void setup_data(El::Int num_neurons, El::Int num_targets, El::Int max_minibatch_size) override {}
+  void set_local_matrix_bypass(CPUMat *m, int idx) override {
+    if(M_local[idx] != nullptr && M_local[idx] != m) {
+      delete M_local[idx];
+    }
+    M_local[idx] = m;
+  }
+  void fp_setup_data(El::Int cur_mini_batch_size, int idx) override {
+    /// @todo BVE FIXME - need to improve how this would work with odd
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/io/data_buffers/partitioned_io_buffer.hpp]: Ok. If we accept that we are introducing a bug, we should open an issue to make it clear.

[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/io/data_buffers/partitioned_io_buffer.hpp]: This has been fixed, although after reviewing the comment and the code, I believe that the comment was addressed by the existing code.  Switching to the Star,VC matrix is a bit cleaner though.

File: scripts/build_lbann_osx.sh [hunk 17]
```diff
@@ -23,7 +23,7 @@ brew_check_install()
 # Check for LBANN dependencies
 brew_check_install git
 brew_check_install cmake
-brew_check_install llvm       # Require OpenMP support in clang
+brew_check_install llvm       # Require OpenMP support in clang brew install --with-toolchain llvm
```
[the remark this answers is not in the document]
Comment (bvanessen) [reply scripts/build_lbann_osx.sh]: I clarified the comment.

File: src/utils/threads/thread_utils.cpp (hunk 13 above)
Comment (bvanessen) [reply src/utils/threads/thread_utils.cpp]: It has been updated to address the Aluminum thread.  The MPI and NCCL threads are not bound to cores and are already a bit out of control.  There is also a command line and prototext option for controlling the I/O threads.

File: include/lbann/base.hpp [hunk 18]
```diff
@@ -163,10 +163,16 @@ static bool __attribute__((used)) endsWith(const std::string mainStr, const std:
 }  // namespace lbann
 
 /// Print the dimensions and name of a Elemental matrix
-static void __attribute__((used)) _display_matrix(ElMat *m, const char *name) {
+static void __attribute__((used)) _print_matrix_dims(AbsDistMat *m, const char *name) {
```
Comment (benson31) [inline include/lbann/base.hpp]: IIRC, `__attribute__` is a GCC extension and not ISO C++. Also, why internal linkage in a file that will be `#include`d in every translation unit in the project and in every translation unit of downstreams?

File: include/lbann/data_readers/data_reader_jag_conduit.hpp [hunk 19]
```diff
@@ -104,6 +104,8 @@ class data_reader_jag_conduit : public generic_data_reader {
   ~data_reader_jag_conduit() override;
   data_reader_jag_conduit* copy() const override { return new data_reader_jag_conduit(*this); }
 
+  void setup(int num_io_threads) override;
```
Comment (benson31) [inline include/lbann/data_readers/data_reader_jag_conduit.hpp]: Doesn't override. Missing the `thread_pool` parameter.

File: include/lbann/data_readers/data_reader_jag_conduit_hdf5.hpp [hunk 20]
```diff
@@ -76,6 +76,8 @@ class data_reader_jag_conduit_hdf5 : public generic_data_reader {
   /// Load data and do data reader's chores.
   void load() override;
 
+  void setup(int num_io_threads) override;
```
Comment (benson31) [inline include/lbann/data_readers/data_reader_jag_conduit_hdf5.hpp]: Doesn't override. Missing `thread_pool` parameter.

File: include/lbann/layers/io/input/generic_input_layer.hpp [hunk 21]
```diff
@@ -848,21 +915,26 @@ class generic_input_layer : public io_layer {
...
 
   data_reader_map_t m_data_readers;
  //  std::map<execution_mode, dataset_stats> m_dataset_stats;
+  bool m_data_set_processed;
+  std::mutex dr_mutex;
 };
 
 template<> inline void generic_input_layer::initialize_io_buffer<partitioned_io_buffer>(lbann_comm *comm, int num_parallel_readers, std::map<execution_mode, generic_data_reader *> data_readers) {
-  io_buffer = new partitioned_io_buffer(comm, num_parallel_readers, data_readers, m_expected_num_child_layers);
+  m_io_buffers.push_back(new partitioned_io_buffer(comm, num_parallel_readers, data_readers, m_expected_num_child_layers));
```
Comment (benson31) [inline include/lbann/layers/io/input/generic_input_layer.hpp]: Why are these specialized? They appear to be the same functions with a different `BufferT` being `new`-ed in each with the same arguments. Am I missing something?

File: include/lbann/utils/threads/type_erased_function.hpp [hunk 22]
```diff
@@ -0,0 +1,91 @@
+#ifndef __LBANN_TYPE_ERASED_FUNCTION_HPP__
+#define __LBANN_TYPE_ERASED_FUNCTION_HPP__
+
+#include <type_traits>
+
+#include "memory.hpp"
+
+namespace lbann {
+
+/** \class type_erased_function
```
Comment (benson31) [inline include/lbann/utils/threads/type_erased_function.hpp]: The documentation of this class is remarkably complete. We should always do that. :P

File: include/lbann/utils/threads/memory.hpp [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <memory>
+
+namespace lbann {
+
+#if __cplusplus < 201402L
+
+/** \brief Local definition of make_unique for non-C++14 compilers */
+template <typename T, typename... Ts>
+std::unique_ptr<T> make_unique(Ts&&... params)
```
Comment (benson31) [inline include/lbann/utils/threads/memory.hpp]: Oh hey, my memory file. Since we have `lbann::make_unique`, we should use it. That should be a different PR, though. @timmoon10

File: src/data_readers/data_reader_imagenet.cpp [hunk 24]
```diff
@@ -43,17 +43,17 @@ imagenet_reader::imagenet_reader(const std::shared_ptr<cv_process>& pp, bool shu
     throw lbann_exception(err.str());
   }
 
-  replicate_processor(*pp);
+  m_master_pps.reset(new cv_process(*pp));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet.cpp]: `m_master_pps = lbann::make_unique<cv_process>(*pp);`

File: src/data_readers/data_reader_imagenet.cpp [hunk 25]
```diff
@@ -43,17 +43,17 @@ imagenet_reader::imagenet_reader(const std::shared_ptr<cv_process>& pp, bool shu
...
 imagenet_reader::imagenet_reader(const imagenet_reader& rhs)
   : image_data_reader(rhs) {
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     std::stringstream err;
     err << __FILE__<<" "<<__LINE__<< " :: " << get_type() << " construction error: no image processor";
     throw lbann_exception(err.str());
   }
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process(*rhs.m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet.cpp]: `m_master_pps = lbann::make_unique<cv_process>(*rhs.m_master_pps);`

File: src/data_readers/data_reader_imagenet.cpp [hunk 26]
```diff
@@ -43,17 +43,17 @@ imagenet_reader::imagenet_reader(const std::shared_ptr<cv_process>& pp, bool shu
...
 }
 
 imagenet_reader& imagenet_reader::operator=(const imagenet_reader& rhs) {
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet.cpp]: Somewhat outside the scope of this PR, but since I'm noticing the terribly duplicated code below: http://gotw.ca/gotw/059.htm @timmoon10

File: src/data_readers/data_reader_imagenet.cpp [hunk 27]
```diff
@@ -64,12 +64,12 @@ imagenet_reader& imagenet_reader::operator=(const imagenet_reader& rhs) {
...
   image_data_reader::operator=(rhs);
 
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     std::stringstream err;
     err << __FILE__<<" "<<__LINE__<< " :: " << get_type() << " construction error: no image processor";
     throw lbann_exception(err.str());
   }
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process(*rhs.m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet.cpp]: `m_master_pps = lbann::make_unique<cv_process>(*rhs.m_master_pps);`

File: src/data_readers/data_reader_imagenet.cpp [hunk 28]
```diff
@@ -84,17 +84,19 @@ void imagenet_reader::set_defaults() {
...
+bool imagenet_reader::replicate_processor(const cv_process& pp, const int nthreads) {
   m_pps.resize(nthreads);
 
   // Construct thread private preprocessing objects out of a shared pointer
-  LBANN_DATA_FETCH_OMP_PARALLEL_FOR_ARGS(schedule(static, 1))
   for (int i = 0; i < nthreads; ++i) {
     //auto ppu = std::make_unique<cv_process>(pp); // c++14
-    std::unique_ptr<cv_process> ppu(new cv_process(pp));
-    m_pps[i] = std::move(ppu);
+    m_pps[i].reset(new cv_process(pp));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet.cpp]: `m_pps[i] = lbann::make_unique<cv_process>(pp);`

File: src/data_readers/data_reader_imagenet_patches.cpp [hunk 29]
```diff
@@ -44,19 +44,19 @@ imagenet_reader_patches::imagenet_reader_patches(const std::shared_ptr<cv_proces
     throw lbann_exception(err.str());
   }
 
-  replicate_processor(*pp);
+  m_master_pps.reset(new cv_process_patches(*pp));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet_patches.cpp]: I'm going to stop marking these. Use `make_unique` here, since we have it.

File: src/data_readers/data_reader_imagenet_patches.cpp [hunk 30]
```diff
@@ -44,19 +44,19 @@ imagenet_reader_patches::imagenet_reader_patches(const std::shared_ptr<cv_proces
...
 {
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     std::stringstream err;
     err << __FILE__<<" "<<__LINE__<< " :: " << get_type() << " construction error: no image processor";
     throw lbann_exception(err.str());
   }
   m_num_patches = rhs.m_num_patches;
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process_patches(*rhs.m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet_patches.cpp]: and here.

File: src/data_readers/data_reader_imagenet_patches.cpp [hunk 31]
```diff
@@ -67,13 +67,13 @@ imagenet_reader_patches& imagenet_reader_patches::operator=(const imagenet_reade
...
 
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     std::stringstream err;
     err << __FILE__<<" "<<__LINE__<< " :: " << get_type() << " construction error: no image processor";
     throw lbann_exception(err.str());
   }
   m_num_patches = rhs.m_num_patches;
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process_patches(*rhs.m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet_patches.cpp]: and here

File: src/data_readers/data_reader_imagenet_patches.cpp [hunk 32]
```diff
@@ -89,17 +89,20 @@ void imagenet_reader_patches::set_defaults() {
...
+bool imagenet_reader_patches::replicate_processor(const cv_process_patches& pp, const int nthreads) {
   m_pps.resize(nthreads);
 
   // Construct thread private preprocessing objects out of a shared pointer
-  LBANN_DATA_FETCH_OMP_PARALLEL_FOR_ARGS(schedule(static, 1))
   for (int i = 0; i < nthreads; ++i) {
     //auto ppu = std::make_unique<cv_process_patches>(pp); // c++14
-    std::unique_ptr<cv_process_patches> ppu(new cv_process_patches(pp));
-    m_pps[i] = std::move(ppu);
+    m_pps[i].reset(new cv_process_patches(pp));
```
Comment (benson31) [inline src/data_readers/data_reader_imagenet_patches.cpp]: and here

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 33]
```diff
@@ -306,7 +306,7 @@ data_reader_jag_conduit::data_reader_jag_conduit(const std::shared_ptr<cv_proces
     _THROW_LBANN_EXCEPTION_(get_type(), " construction error: no image processor");
   }
 
-  replicate_processor(*pp);
+  m_master_pps.reset(new cv_process(*pp));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit.cpp]: and here

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 34]
```diff
@@ -325,11 +325,11 @@ void data_reader_jag_conduit::copy_members(const data_reader_jag_conduit& rhs) {
   m_scalar_keys = rhs.m_scalar_keys;
   m_input_keys = rhs.m_input_keys;
 
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     _THROW_LBANN_EXCEPTION_(get_type(), " construction error: no image processor");
   }
 
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process(*m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit.cpp]: and here

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 35]
```diff
@@ -417,17 +417,19 @@ void data_reader_jag_conduit::set_defaults() {
...
+bool data_reader_jag_conduit::replicate_processor(const cv_process& pp, const int nthreads) {
   m_pps.resize(nthreads);
 
   // Construct thread private preprocessing objects out of a shared pointer
-  LBANN_DATA_FETCH_OMP_PARALLEL_FOR_ARGS(schedule(static, 1))
   for (int i = 0; i < nthreads; ++i) {
     //auto ppu = std::make_unique<cv_process>(pp); // c++14
-    std::unique_ptr<cv_process> ppu(new cv_process(pp));
-    m_pps[i] = std::move(ppu);
+    m_pps[i].reset(new cv_process(pp));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit.cpp]: and here

File: src/data_readers/data_reader_jag_conduit_hdf5.cpp [hunk 36]
```diff
@@ -73,7 +73,7 @@ data_reader_jag_conduit_hdf5::data_reader_jag_conduit_hdf5(const std::shared_ptr
     _THROW_LBANN_EXCEPTION_(get_type(), " construction error: no image processor");
   }
 
-  replicate_processor(*pp);
+  m_master_pps.reset(new cv_process(*pp));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit_hdf5.cpp]: and here

File: src/data_readers/data_reader_jag_conduit_hdf5.cpp [hunk 37]
```diff
@@ -87,11 +87,11 @@ void data_reader_jag_conduit_hdf5::copy_members(const data_reader_jag_conduit_hd
   m_input_keys = rhs.m_input_keys;
   m_success_map = rhs.m_success_map;
 
-  if (rhs.m_pps.size() == 0u || !rhs.m_pps[0]) {
+  if (!rhs.m_master_pps) {
     _THROW_LBANN_EXCEPTION_(get_type(), " construction error: no image processor");
   }
 
-  replicate_processor(*rhs.m_pps[0]);
+  m_master_pps.reset(new cv_process(*rhs.m_master_pps));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit_hdf5.cpp]: and here

File: src/data_readers/data_reader_jag_conduit_hdf5.cpp [hunk 38]
```diff
@@ -127,16 +127,18 @@ void data_reader_jag_conduit_hdf5::set_defaults() {
...
+/// Replicate image processor for each I/O thread
+  bool data_reader_jag_conduit_hdf5::replicate_processor(const cv_process& pp, const int nthreads) {
   m_pps.resize(nthreads);
 
   // Construct thread private preprocessing objects out of a shared pointer
-  LBANN_DATA_FETCH_OMP_PARALLEL_FOR_ARGS(schedule(static, 1))
   for (int i = 0; i < nthreads; ++i) {
-    std::unique_ptr<cv_process> ppu(new cv_process(pp));
-    m_pps[i] = std::move(ppu);
+    m_pps[i].reset(new cv_process(pp));
```
Comment (benson31) [inline src/data_readers/data_reader_jag_conduit_hdf5.cpp]: and here

File: src/data_readers/data_reader_merge_features.cpp [hunk 39]
```diff
@@ -43,7 +43,7 @@ data_reader_merge_features::data_reader_merge_features(
   const data_reader_merge_features& other) :
   generic_compound_data_reader(other),
   m_data_size(other.m_data_size) {
-  if(other.m_label_reader != nullptr) 
+  if(other.m_label_reader != nullptr)
```
Comment (benson31) [inline src/data_readers/data_reader_merge_features.cpp]: pointers implicitly cast to `bool`? Why `!= nullptr`?

File: src/io/data_buffers/partitioned_io_buffer.cpp [hunk 40]
```diff
@@ -29,48 +29,97 @@
...
 lbann::partitioned_io_buffer::partitioned_io_buffer(lbann_comm *comm, int num_parallel_readers, std::map<execution_mode, generic_data_reader *> data_readers, int num_child_layers)
   : generic_io_buffer(comm, num_parallel_readers, data_readers),
-    M_local(num_child_layers, nullptr) {}
+    m_num_samples_fetched(0) {
+  if (num_child_layers != 2) { /// @todo Generalize
+    LBANN_ERROR("currently only supported with two child layers");
+  }
+  m_input_buffers.clear();
+  m_input_buffers.resize(num_child_layers);
+  m_input_buffers[0].reset(AbsDistMat::Instantiate(comm->get_model_grid(), 0,
```
Comment (benson31) [inline src/io/data_buffers/partitioned_io_buffer.cpp]: I wouldn't use `Instantiate` here. That function is for dynamic construction and this should clearly be static. The type you want is `DistMatrix<DataType, El::Dist::STAR, El::Dist::VC, El::DistWrap::Element, El::Device::CPU>` and I statically know that because you typed it. This will save you a terrifying abhorrent `if ... else if ... else if ... else if ... else if ... else if ... else if ... *else if ... * <repeat * to * ad nauseum> else` block.

File: src/io/data_buffers/partitioned_io_buffer.cpp [hunk 41]
```diff
@@ -29,48 +29,97 @@
...
+                            El::STAR, El::VC, El::ELEMENT, El::Device::CPU));
+  m_input_buffers[1].reset(AbsDistMat::Instantiate(comm->get_model_grid(), 0,
```
Comment (benson31) [inline src/io/data_buffers/partitioned_io_buffer.cpp]: Ditto

File: src/utils/threads/thread_pool.cpp [hunk 42]
```diff
@@ -0,0 +1,161 @@
...
+void thread_pool::launch_threads(size_type num_threads)
+{
+  threads_.reserve(num_threads);
+
+  // Try to launch each worker thread
+  try
+  {
+    for (size_type cnt = 0; cnt < num_threads; ++cnt) {
+      threads_.emplace_back(&thread_pool::do_thread_work_,this);
+      // pthread_t my_thread_native = threads_[cnt].native_handle();
```
Comment (benson31) [inline src/utils/threads/thread_pool.cpp]: Is this code possibly going to be resurrected? If not (and even if so), perhaps it should be deleted.

File: src/utils/threads/thread_pool.cpp [hunk 43]
```diff
@@ -0,0 +1,161 @@
...
+      CPU_ZERO(&ht_cpuset);
+      // Pin this thread to the base CPU id plus the thread count and offset
+      for (int j = 0; j < CPU_SETSIZE; j++) {
+        if (CPU_ISSET(j, &cpuset)) {
+          CPU_SET(j+cnt+cpu_offset, &ht_cpuset);
+        }
+      }
+
+      threads_.emplace_back(&thread_pool::do_thread_work_pinned_thread_,this, cnt, ht_cpuset);
+      // pthread_t my_thread_native = threads_[cnt].native_handle();
```
Comment (benson31) [inline src/utils/threads/thread_pool.cpp]: Ditto

File: include/lbann/utils/threads/memory.hpp (hunk 23 above)
Comment (timmoon10) [reply include/lbann/utils/threads/memory.hpp]: Nifty. Pinging Issue #155.

File: src/data_readers/data_reader_merge_features.cpp (hunk 39 above)
Comment (timmoon10) [reply src/data_readers/data_reader_merge_features.cpp]: We use the `if (ptr != nullptr)` idiom throughout the codebase. I think it's marginally clearer, but perhaps it is excessive.

Comment (benson31) [reply src/data_readers/data_reader_merge_features.cpp]: Eh, it's not a big deal. It just stood out as overly verbose. For folks not using IDEs, I suppose it suggests the pointer nature of `m_label_reader`, which is good documentation. I rescind my complaint.

File: include/lbann/data_readers/data_reader_jag_conduit.hpp (hunk 19 above)
Comment (bvanessen) [reply include/lbann/data_readers/data_reader_jag_conduit.hpp]: Fixed.

File: include/lbann/layers/io/input/generic_input_layer.hpp (hunk 21 above)
Comment (bvanessen) [reply include/lbann/layers/io/input/generic_input_layer.hpp]: Because the I/O buffer is very different between the partitioned and distributed versions.  Also the distributed I/O buffer is going to be deprecated soon.

File: include/lbann/utils/threads/type_erased_function.hpp (hunk 22 above)
Comment (bvanessen) [reply include/lbann/utils/threads/type_erased_function.hpp]: :-D

File: include/lbann/layers/io/input/generic_input_layer.hpp (hunk 21 above)
Comment (benson31) [reply include/lbann/layers/io/input/generic_input_layer.hpp]: That's fine, but I mean "why is the source code duplicated when all that differs is the type (but not the arguments)?" Should this and the next function not just be collapsed into the single function template:
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, data_readers, m_expected_num_child_layers});
}
```

Comment (benson31) [reply include/lbann/layers/io/input/generic_input_layer.hpp]: Also, do you mean to do a deep copy of the `std::map` object there? You're doing at least 2, one when you enter `initialize_io_buffer` and one when in the ctor of your `{partitioned|distributed}_io_buffer`. If you need a clean copy to go into the buffer, the function should *actually* be
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, std::move(data_readers), m_expected_num_child_layers});
}
```
so that you only do the one copy that you need and then move the clean copy into the buffer. If, on the other hand, you don't know if the buffer needs a clean copy, the function should be written:
```cpp
template <typename T> inline void generic_input_layer::initialize_io_buffer(
    lbann_comm *comm, int num_parallel_readers,
    std::map<execution_mode, generic_data_reader *> const& data_readers) {
  m_io_buffers.push_back(new T{comm, num_parallel_readers, data_readers, m_expected_num_child_layers});
}
```
so that the buffer constructor can decide for itself if it needs a clean copy and it is then responsible for constructing it.

File: src/io/data_buffers/partitioned_io_buffer.cpp (hunk 40 above)
Comment (bvanessen) [reply src/io/data_buffers/partitioned_io_buffer.cpp]: Fixed.

File: include/lbann/layers/io/input/generic_input_layer.hpp (hunk 21 above)
Comment (bvanessen) [reply include/lbann/layers/io/input/generic_input_layer.hpp]: I think that we can fix this up after some code cleanup in a future PR.

**Comment by benson31** [comment]
Glad this merged finally! 👍
